src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -406,6 +402,39 @@ void InitializeCurrentProcessCpuCount()
     g_RhNumberOfProcessors = count;
 }
 
+typedef void (*Destructor) (void*);
+typedef int (*AtexitFunc) (Destructor, void*, void*);
+
+#ifdef TARGET_LINUX
+static pthread_key_t key;
+static pthread_once_t once = PTHREAD_ONCE_INIT;
+
+struct Link
+{
+    Link* Next;
+    Destructor Dtor;
+    void* Arg;
+};
+
+static void CallDestructors(void* obj)
+{
+    Link* next;
+    Link* link = (Link*)obj;
+    while (link)
+    {
+        next = link->Next;
+        link->Dtor(link->Arg);
+        free(link);
+        link = next;
+    }
+}
+
+static void CreateKey(void)
+{
+    pthread_key_create(&key, CallDestructors);
```

Comment:
```suggestion
    pthread_key_create(&key, RuntimeThreadShutdown);
```

Reply:
I have cleaned this up. This refactoring is working after moving the key creation to PalInit. When we were using `__cxa_thread_exit`, linked list was required then. Thanks for pushing it. :)

(waiting for CI to see how `-fno-threadsafe-statics` pans out, locally osx and linux were happy)